### PR TITLE
[REF] make `TestCursor` use the test case cursor

### PR DESCRIPTION
**Pull request by rco-odoo**
This allows rpc requests in `HttpCase` to use the cursor `self.cr`, which is now shared between the Python test and the rpc requests.  This simplifies code to prepare a JS test, and code to check the result of a JS tour.

### CODE REVIEW

File: odoo/sql_db.py [hunk 1]
```diff
@@ -425,42 +426,72 @@ def __getattr__(self, name):
...
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is None:
+            self.commit()
+        self.close()
+
+    def __getattr__(self, name):
+        value = getattr(self._cursor, name)
+        if callable(value) and self._closed:
+            raise psycopg2.OperationalError('Unable to use a closed cursor.')
```
Comment (kujiu) [inline odoo/sql_db.py]: Add the value on message when raising an error.

File: odoo/sql_db.py [hunk 2]
```diff
@@ -425,42 +426,72 @@ def __getattr__(self, name):
...
-        self.execute("RELEASE SAVEPOINT test_cursor")
-        self.execute("SAVEPOINT test_cursor")
+        self._cursor.execute('RELEASE SAVEPOINT "%s"' % self._savepoint)
+        self._cursor.execute('SAVEPOINT "%s"' % self._savepoint)
 
     def rollback(self):
-        self.execute("ROLLBACK TO SAVEPOINT test_cursor")
-        self.execute("SAVEPOINT test_cursor")
+        self._cursor.execute('ROLLBACK TO SAVEPOINT "%s"' % self._savepoint)
+        self._cursor.execute('SAVEPOINT "%s"' % self._savepoint)
```
Comment (odony) [inline odoo/sql_db.py]: this line is not really necessary

File: odoo/sql_db.py [hunk 3]
```diff
@@ -425,42 +426,72 @@ def __getattr__(self, name):
...
 
-    def force_close(self):
-        super(TestCursor, self).close()
+    def __init__(self, cursor, lock):
+        self._closed = False
+        self._cursor = cursor
+        self._lock = lock
+        self._savepoint = "test_cursor_%s" % next(self._savepoint_seq)
+        # we use a lock to serialize concurrent requests
+        self._lock.acquire()
```
Comment (odony) [inline odoo/sql_db.py]: swap with previous line to protect the sequence allocation

**Comment by sle-odoo** [comment]
fixes #12237

**Pull request merged**